test/e2e/sagemaker/tests/test_hpo.py

```diff
@@ -0,0 +1,151 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	 http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+"""Integration tests for the SageMaker TrainingJob API.
+"""
+
+import boto3
+import pytest
+import logging
+from typing import Dict
+import time
+
+from sagemaker import SERVICE_NAME, service_marker, CRD_GROUP, CRD_VERSION
+from sagemaker.replacement_values import REPLACEMENT_VALUES
+from common.resources import load_resource_file, random_suffix_name
+from common import k8s
+
+RESOURCE_PLURAL = 'hyperparametertuningjobs'
+
+
+@pytest.fixture(scope="module")
+def sagemaker_client():
+    return boto3.client('sagemaker')
+
+
+@pytest.fixture(scope="module")
+def xgboost_hpojob():
+    resource_name = random_suffix_name("xgboost-hpojob", 32)
+
+    replacements = REPLACEMENT_VALUES.copy()
+    replacements["HPO_JOB_NAME"] = resource_name
+
+    hpojob = load_resource_file(
+        SERVICE_NAME, "xgboost_hpojob", additional_replacements=replacements)
+    logging.debug(hpojob)
+
+    # Create the k8s resource
+    reference = k8s.CustomResourceReference(
+        CRD_GROUP, CRD_VERSION, RESOURCE_PLURAL, resource_name, namespace="default")
+    resource = k8s.create_custom_resource(reference, hpojob)
+    resource = k8s.wait_resource_consumed_by_controller(reference)
+
+    assert resource is not None
+
+    yield (reference, resource)    
+
+    # Delete the k8s resource if not already deleted by tests
+    try:
+        k8s.delete_custom_resource(reference)
```

Comment:
This is deleted in one of the tests below, which actually indicates that this should not actually be a fixture (which is some test dependency that should be set up and torn down automatically by pytest) but rather something that should be created specifically for a test of the create/delete functionality as a whole (see more below)

Reply:
Intention is, even if the test fails, we clean up all the resources which cost $$. So I think author started using fixture rather than putting try catch blocks in the test method.

is this not good?

Reply:
> (edit) now that I think about the above statement, script should be throwing error if delete fails rather than ignoring the exception because it means we have a leaking resource

Yes, exactly.

Though, my real point here is that we should be considering what the thing is that we are testing and what are those test dependencies.

What I mean by that is when testing the Create [-> Update] -> Delete code paths for objects of type X that we are not using fixtures that call the Create, Update or Delete code paths, since it is those code paths that are the exact part of the system under test that we wish to assert.

See https://docs.pytest.org/en/latest/fixture.html#what-fixtures-are for more details on this concept. There's nothing wrong with having a fixture that creates, yields and cleans up an object of type X. However, such fixtures should be dependencies of tests that are asserting behaviour and actions against objects NOT of type X.

In other words, you should feel free to make a little helper function like this:

```python
def _make_hpojob():
    resource_name = random_suffix_name("xgboost-hpojob", 32)

    replacements = REPLACEMENT_VALUES.copy()
    replacements["HPO_JOB_NAME"] = resource_name

    data = load_resource_file(
        SERVICE_NAME, "xgboost_hpojob", additional_replacements=replacements)
    logging.debug(data)

    return k8s.CustomResourceReference(
        CRD_GROUP, CRD_VERSION, RESOURCE_PLURAL, resource_name, namespace="default")
```

and call this helper function from a fixture that creates, yields and deletes an HPO Job:

```python
@pytest.fixture(scope="module")
def xgboost_hpojob():
    hpo_job = _make_hpojob()
    resource = k8s.create_custom_resource(hpo_job)
    resource = k8s.wait_resource_consumed_by_controller(hpo_job)

    yield (hpo_job, resource)    

    k8s.delete_custom_resource(hpo_job)
```

(note the above has no error suppression around the delete call since we definitely want to error out if deletion doesn't succeed or the resource has already been deleted, because for fixtures, our expectation is that the *fixture* controls the lifecycle of the object being yielded by the fixture, not the caller.

And then, to actually test the Create and Delete actions for an HPO Job, you'd want a test method that runs, in a serial fashion, the same create/delete calls along with assertions about state transitions:

```python
@service_marker
@pytest.mark.canary
class TestHPO:
    def test_smoke(self):
        hpo_job = _make_hpojob()
        resource = k8s.create_custom_resource(reference, hpo_job)
        resource = k8s.wait_resource_consumed_by_controller(reference)
        assert k8s.get_resource_exists(hpo_job)

        hpo_job_name = resource["spec"].get("hyperParameterTuningJobName", None)
        assert hpo_job_name is not None

        hpo_job = get_sagemaker_hpo_job(hpo_job_name)

        assert k8s.get_resource_arn(resource) == hpo_job["HyperParameterTuningJobArn"]
        assert hpo_job["HyperParameterTuningJobStatus"] in HPO_JOB_STATUS_CREATED

        _, deleted = k8s.delete_custom_resource(reference)
        assert deleted is True

        hpo_job = get_sagemaker_hpo_job(hpo_job_name)
        assert hpo_job["HyperParameterTuningJobStatus"] in HPO_JOB_STATUS_STOPPED
```

Now, let's say you had a test for the create and delete code paths for a **different** object of type Y and objects of type Y required an HPO Job to already exist (because, say, objects of type Y referred to an HPO Job by ID).

This is the case where you'd want to use your `xgboost_hpojob` fixture, because the create and delete code paths of the HPO Job itself are **not** the thing being tested. Instead, the create and delete code paths of Object of type Y is what is being tested. So, you'd use the `xgboost_hpojob` as a fixture in the method that tested the create and delete code paths of Objects of type Y, like so:

```python
@service_marker
@pytest.mark.canary
class TestY:
    def test_smoke(self, xgboost_hpojob):
        hpo_job_ref, _ = xgboost_hpojob
        hpo_job_name = hpo_job_ref["spec"].get("hyperParameterTuningJobName"]

        object_y = _make_object_y(hpo_job_name)
        resource = k8s.create_custom_resource(object_y)
        resource = k8s.wait_resource_consumed_by_controller(object_y)
        assert k8s.get_resource_exists(object_y)

        # bunch of assertions...

        _, deleted = k8s.delete_custom_resource(object_y)
        assert deleted is True

        # bunch of assertions... but do NOT delete the HPO Job fixture,
        # since the fixture owns the lifecycle of the HPO Job and not
        # this `TestY.test_smoke` method.
```

Reply:
oh this is neat 

In the `xgboost_hpojob` fixture we only need to  delete it if it exists else it will throw a "resource not found" error. 
This is required because some tests might delete the resource during testing.

--------

By combining these tests into 1, we lose a few thing :
- While running locally we have to wait for whole thing to pass/fail and then we see the logs. Instead of seeing each test pass/fail as we go.
- Now that we only have 1 test, it can fail because of multiple reasons and we won't be able to track that metric.

--------------

Let me know what you guys think 
Once we merge this PR i can make this same change for all of our tests

Reply:
> oh this is neat
> 
> In the `xgboost_hpojob` fixture we only need to delete it if it exists else it will throw a "resource not found" error.
> This is required because some tests might delete the resource during testing.

There should not be assertions in fixtures. Fixtures are for test dependencies. Fixtures should handle the lifecycle of these test dependencies entirely. If a test deletes an object that is `yield`'d from a fixture and that fixture deletes the object after the `yield`, that is an indication that the **test should not be using that fixture**, because the `yield`'d object is no longer a test *dependency*, rather, the object is the *target* of the test's Actions and Assertions.

> By combining these tests into 1, we lose a few thing :
> 
>     * While running locally we have to wait for whole thing to pass/fail and then we see the logs. Instead of seeing each test pass/fail as we go.

The `assert` statements will fail-fast the test. Once something fails, the test will stop running. This is actually what you want, no? Why would you want to continue running a test about a Status attribute if the `create` call failed?

>     * Now that we only have 1 test, it can fail because of multiple reasons and we won't be able to track that metric.

What metric are you referring to here?

Reply:
In the example code `test_smoke()` 👆🏼
```
_, deleted = k8s.delete_custom_resource(object_y)
assert deleted is True
```
This will delete the resource. So this test should not be using the `xgboost_hpojob` fixture right?

-------------------
Combining the tests : 
- yes we will see the assert failures. I meant, while running tests locally if assertions are passing we will not know what stage the test is at. We will only see the logs once the whole thing is done which can take a while. (Eg: SageMaker Endpoint)
- Metrics : I meant like how long did delete test take ? or how many times update failed over time ?. 
By having create/update/delete in one test we will not be able to track individual metrics.

Reply:
> In the example code `test_smoke()` 👆🏼
> 
> ```
> _, deleted = k8s.delete_custom_resource(object_y)
> assert deleted is True
> ```
> 
> This will delete the resource. So this test should not be using the `xgboost_hpojob` fixture right?

Correct.

> Combining the tests :
> 
>     * yes we will see the assert failures. I meant, while running tests locally if assertions are passing we will not know what stage the test is at. We will only see the logs once the whole thing is done which can take a while. (Eg: SageMaker Endpoint)

I guess I don't see this as a problem.

>     * Metrics : I meant like how long did delete test take ? or how many times update failed over time ?.

I don't see these tests as the source of metrics for things like how long a delete of a resource takes. A couple points about metrics like this:
1) That is essentially a server-side metric that the Sagemaker service should be responsible for tracking, since the length of time to delete of a resource should not be dependent on the client that calls the delete operation.
2) What you would be tracking here is essentially the *poll time*, not the actual delete time. For the actual delete time, only the server side knows that.

For things like how many times an update failed over time, I presume you are talking about how many times the update code path *in a specific test failed over time* (i.e. a regression test analysis)? If you're talking about how many times a particular Update operation failed on the server side, please see point 1) above.

For regression analysis I generally prefer unit tests which can be an effective way to test individual parts of a functional test's code paths by mocking out test dependencies and isolating just one part of the code base under test. Functional integration or end-to-end tests, however, are meant to verify the full behaviour of the system under test. What these tests are trying to do is validate specific use cases in their entirety. The use case being validated here is a "smoke test" of the Create Resource -> Read Resource -> Delete Resource use case. Other use cases I hope that e2e tests will validate are things like:

* Adopting a pre-existing resource
* Cross-account resource management
* Upgrading the version of a controller and verifying resources under management can still be managed

Reply:
Hey @jaypipes,

So I've just taken this advice listed above for work with the new Application Auto Scaling integration tests I've been working on lately. I think the only thing missing from your proposed solution is a way of correctly handling cleanup of tests. The great thing about the fixtures is the ability to `yield` and delete regardless of the assertions within a test. 

As soon as you handle the CRUD lifecycle of the CR within the test body, if you run into a failed assertion prior to the delete instruction then the test will fail early and never clean up its own resources. This then becomes a problem when attempting to retry tests, as older cluster resources could collide with the new ones.

One such proposed solution would be to have a cleanup fixture, with which you can register all of your custom resource references in each test, that will ensure that the tests at least call `delete` before exiting. 

Do you think I'm missing some obvious solution that otherwise could solve this issue?

Reply:
Sorry @RedbackThomson I seem to have missed your comment above, apologies!

> As soon as you handle the CRUD lifecycle of the CR within the test body, if you run into a failed assertion prior to the delete instruction then the test will fail early and never clean up its own resources. This then becomes a problem when attempting to retry tests, as older cluster resources could collide with the new ones.

I guess I don't see this as a problem, because if an assertion about a resource under test fails, typically we *want* the resource to stay around so we can diagnose *why* the assertion failed. Note that I am specifically referring to resources *under test* -- meaning, the specific resource we are asserting about. Fixtures (and the yield/teardown functionality) are for resources that are *dependencies* of tests, not resources that are themselves the subject of assertions.

Reply:
I would agree up until your CI system runs the tests 30 times a day, each of them failing because of a poorly written test, and then you start to hit resource quotas and limits in the AWS account. This is what we saw happening in our long-running canary tests for SageMaker and it required a separate script that cleaned up hanging resources. It isn't ideal, because we have to manage that script separately from our tests.

Reply:
> I would agree up until your CI system runs the tests 30 times a day, each of them failing because of a poorly written test, and then you start to hit resource quotas and limits in the AWS account. This is what we saw happening in our long-running canary tests for SageMaker and it required a separate script that cleaned up hanging resources. It isn't ideal, because we have to manage that script separately from our tests.

This is actually how upstream Kubernetes CI system works. There is a separate process called Boskos that acts as a janitor for the Prow CI jobs and ensures resources are cleaned up:

https://github.com/kubernetes-sigs/boskos

Eventually, when we move to a Prow-based system, we might consider moving our bootstrap/cleanup for some things into Boskos configurations.